src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets

```diff
@@ -151,6 +154,75 @@
     </ItemGroup>
   </Target>
 
+  <!-- #################################### -->
+  <!-- Generate BuildVersion.props -->
+  <!-- #################################### -->
+  <Target Name="CreateOrUpdateCurrentVersionFile" Condition="'$(SkipVersionGeneration)'!='true'" DependsOnTargets="GenerateCurrentVersion">
+    <Error Condition="'$(OverrideBuildNumberMajor)'==''" Text="Error when generating current version." />
+
+    <ItemGroup>
+      <CurrentVersionLines Include="%3C%3Fxml version=%221.0%22 encoding=%22utf-8%22%3F%3E" />
+      <CurrentVersionLines Include="%3C!-- This is a generated file. To re-generate it, run build.cmd --%3E" />
+      <CurrentVersionLines Include="%3CProject xmlns=%22http://schemas.microsoft.com/developer/msbuild/2003%22%3E" />
+      <CurrentVersionLines Include="%3CPropertyGroup%3E" />
+      <CurrentVersionLines Include="%3CBuildNumberMajor Condition=%22%27%24(BuildNumberMajor)%27==%27%27%22%3E$(OverrideBuildNumberMajor)%3C/BuildNumberMajor%3E" />
+      <CurrentVersionLines Include="%3C/PropertyGroup%3E" />
+      <CurrentVersionLines Include="%3C/Project%3E" />
+      <BuildVersionFileItem Include="$(BuildVersionFile)" />
+    </ItemGroup>
+
+    <PropertyGroup>
+      <BuildVersionSemaphoreFile>%(BuildVersionFileItem.RootDir)%(BuildVersionFileItem.Directory)semaphore.txt</BuildVersionSemaphoreFile>
+      <ShouldLockFileWrite Condition="!Exists('$(BuildVersionSemaphoreFile)')">true</ShouldLockFileWrite>
+    </PropertyGroup>
+
+    <MakeDir Condition="!Exists('%(BuildVersionFileItem.RootDir)%(BuildVersionFileItem.Directory)')" Directories="%(BuildVersionFileItem.RootDir)%(BuildVersionFileItem.Directory)" />
+
+    <Touch AlwaysCreate="true" Files="$(BuildVersionSemaphoreFile)" Condition="'$(ShouldLockFileWrite)'=='true'" />
+
+    <WriteLinesToFile
+      ContinueOnError="WarnAndContinue"
+      Condition="'$(ShouldLockFileWrite)'=='true'"
+      File="$(BuildVersionFile)"
+      Lines="@(CurrentVersionLines)"
+      Overwrite="true" />
+
+  </Target>
+
+  <Target Name="GenerateCurrentVersion" DependsOnTargets="GetLatestCommitDate">
+    <Error Condition="'$(DateToUse)'==''" Text="A date is needed to Generate the current version." />
+
+    <GenerateCurrentVersion SeedDate="$(DateToUse)">
+      <Output PropertyName="OverrideBuildNumberMajor" TaskParameter="GeneratedVersion" />
+    </GenerateCurrentVersion>
+  </Target>
+
+  <Target Name="GetLatestCommitDate" Condition="'$(DateToUse)'==''">
+    <Exec Command="git show -s --format=^%25%25cd --date=short HEAD" IgnoreExitCode="true" IgnoreStandardErrorWarningFormat="true" ConsoleToMSBuild="true">
```

Comment:
We need this because of DOS escaping, the reason is that the exec task secretly runs everything in a shell instead of directly calling createprocess, so an extra escape is required. I still have to test this in Mac to make sure that this also works there.